lib/src/state_machine_controller.dart

```diff
@@ -1,17 +1,51 @@
+import 'package:flutter/foundation.dart';
+import 'package:rive/src/core/core.dart';
+import 'package:rive/src/generated/animation/state_machine_bool_base.dart';
+import 'package:rive/src/generated/animation/state_machine_number_base.dart';
+import 'package:rive/src/generated/animation/state_machine_trigger_base.dart';
 import 'package:rive/src/rive_core/animation/state_machine.dart';
+import 'package:rive/src/rive_core/animation/state_machine_bool.dart';
+import 'package:rive/src/rive_core/animation/state_machine_input.dart' as core;
+import 'package:rive/src/rive_core/animation/state_machine_number.dart';
+import 'package:rive/src/rive_core/animation/state_machine_trigger.dart';
 import 'package:rive/src/rive_core/artboard.dart';
 import 'package:rive/src/rive_core/state_machine_controller.dart' as core;
 
-class StateMachineInput<T> {
-  final int id;
+/// [StateMachine]s supports three input types. The StateMachine mostly
+/// abstracts types by allowing the programmer to query for an input of a
+/// specific Dart backing type, mapping it to the correct StateMachine type.
+/// This is the most flexible API to use to check if a type with a given name
+/// exists. However, if you need to iterate inputs and query their types, this
+/// enum is exposed for convenience.
+enum SMIType { number, boolean, trigger }
+
+/// StateMachine Instance Input. This is the abstraction of an instanced input
+/// from the [StateMachine]. Whenever a [StateMachineController] is created, the
+/// list of inputs in the corresponding [StateMachine] is wrapped into a set of
+/// [SMIInput] objects that ensure inputs are initialized to design-time values.
+/// The implementation can now change these values freely as they are decoupled
+/// from the backing [StateMachine] and can safely be re-instanced by another
+/// controller later.
+class SMIInput<T> {
+  final core.StateMachineInput _input;
   final StateMachineController controller;
-  StateMachineInput._(this.id, this.controller);
+  final SMIType type;
 
-  T get value => controller.inputValues[id] as T;
-  set value(T newValue) => change(newValue);
+  SMIInput._(this._input, this.type, this.controller);
+
+  @protected
+  void advance() {}
 
-  /// Change the value of the input, returns true if the value was changed the
-  /// and [StateMachineController] was activated.
+  /// The id of the input within the context of the [StateMachine] it belongs
+  /// to.
+  int get id => _input.id;
+
+  /// The name given to this input at design time in Rive.
+  String get name => _input.name;
+
+  /// Convenience method for changing the backing value of the input. Usually
+  /// it's easier to use the various value getter/setters on the derived
+  /// version of [SMIInput] like [SMIBool], [SMINumber], and [SMITrigger].
```

Comment:
Are you referring only to `SMITrigger.fire` in this case or am I missing a member?

Reply:
If you mean the `value` setter on `SMIInput`, I think you can link to `[value]` directly.